### DISCUSSION TITLE
Should it be possible to drop into the R console from the nix-shell and build script within that environment?

### DISCUSSION THREAD

**Discussion by oTheAnalyst** [ropensci/rix#182] [General]
Hello Everyone!

I have had some fun running this package, great work so far! 

One thing I'm having some difficulty with is running R code through the console while in the shell environment. I can run my script through the command line with no issues, and I believe the utility is there for locking in analysis to a specific version of R after the whole script is written and debugged. 

[here is the script I have been testing](https://github.com/VtheRtech/trade-union-behaviour-2020s) 

[Here is neovim configurations I am using along with associated .Rprofile in the readme](https://github.com/VtheRtech/Vim-Config-Files.git)

I am curious is this mainly an issue with the environment I'm running? or perhaps I should be running the `env_build.R` script every time to enter the environment that was built using nix within the R console?

Below is better visual show case of what I am trying to do:

![image](https://github.com/b-rodrigues/rix/assets/30744769/3bb11a84-69e7-4635-8102-00e8e82a6b6c)

When I try going to my text editor through the nix-shell these errors appear the environment doesn't seem to work. R doesn't see `{targets}` or any other relevant package installed. 

```
before_rns.R stderr: 
.build_packages() exit status 0
* installing to library ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’
Error: ERROR: no permission to install to directory ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’
Error in do_exit(status = status) : .install_packages() exit status 1Execution halted

before_rns.R stdout: * checking for file ‘/home/pretender/.local/share/nvim/lazy/R.nvim/nvimcom/DESCRIPTION’ ...
 OK
* preparing ‘nvimcom’:
* checking DESCRIPTION meta-information ...
 OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘nvimcom_0.9.41.tar.gz’

Time:
  global setup: 0.005199992
  before_rns.R (async): 0.394760645
  check health (async): 0.001037178
  unix setup: 3.0948e-05
```

```
Client r_language_server quit with exit code 1 and signal 0. Check log for errors: /home/pretender/.local/state/nvim/lsp.log
```

**Comment by b-rodrigues** [comment]
Do you have an `install.packages()` in your .Rprofile? that seems to be the case and it's not going to work from a nix shell. These packages should be installed with Nix instead. Then you should remove these calls from the .Rprofile and it should work

**Reply by oTheAnalyst** [reply]
I do not seem to have `install.packages()` in my .Rprofile

would this be an issue in the .Rprofile: 

```r
.libPaths(c(.libPaths(), "~/Rpackages/"))
.libPaths(c(.libPaths(), "~/R/x86_64-pc-linux-gnu-library/4.3/"))
```

**Reply by b-rodrigues** [reply]
Can you post your entire .Rprofile? Something seems to try to install a package. Adding this package to the environment should likely solve the problem I think. Regarding the libpaths, I think that could also cause issues potentially but only with subshells and only if you don't use `rix_init()` to generate a .Rprofile that should be loaded by these subshells

**Reply by oTheAnalyst** [reply]
```r

options(width = 100)
.libPaths(c(.libPaths(), "~/Rpackages/"))
.libPaths(c(.libPaths(), "~/R/x86_64-pc-linux-gnu-library/4.3/"))

if (interactive()) {
  view_xl <- function(.data) {
    tmp <- tempfile(fileext = ".csv")
    data.table::fwrite(.data, tmp)
    browseURL(tmp, browser = "gnumeric")
  }

  httpgd::hgd()

  options(browser = "brave",
    reactable.theme = reactable::reactableTheme(
      color = "hsl(233, 9%, 87%)",
      backgroundColor = "#002b36",
      borderColor = "#eee8d5",
      stripedColor = "#586e75",
      highlightColor = "#6c71c4",
      inputStyle = list(backgroundColor = "hsl(233, 9%, 25%)"),
      selectStyle = list(backgroundColor = "hsl(233, 9%, 25%)"),
      pageButtonHoverStyle = list(backgroundColor = "hsl(233, 9%, 25%)"),
      pageButtonActiveStyle = list(backgroundColor = "hsl(233, 9%, 28%)")
  ))

  view <- function(.data) {
    # Use a temporary HTML file to display the table
    tmp <- tempfile(fileext = ".html")

    # Write the reactable output to the temporary HTML file
    html <- reactable::reactable(.data,
      filterable = TRUE,
      searchable = TRUE,
      showPageSizeOptions = TRUE,
      striped = TRUE,
      highlight = TRUE,
      compact = TRUE,
      defaultPageSize = 30
    )
    htmlwidgets::saveWidget(html, file = tmp)

    # Open the temporary HTML file in Brave browser
    # Replace "firefox" with the path to the firefox executable if necessary
    browseURL(tmp, browser = "brave")
  }
}

print(".Rprofile loaded successfully")


```

sure thing!

**Comment by b-rodrigues** [comment]
you might need to install the languageserver package for neovim

**Reply by oTheAnalyst** [reply]
[I double-checked it seems this was already installed.](https://github.com/VtheRtech/Vim-Config-Files/blob/main/lua/config/nvim-lspconfig.lua)

**Comment by oTheAnalyst** [comment]
Thanks, for the suggestion I will take a look!

**Comment by b-rodrigues** [comment]
I see some require() in that config you shared wonder if that's the issue. Try to drop in a shell with a vanilla nvim session, it might be lazyvim that messes things up

**Reply by oTheAnalyst** [reply]
I must confess I don't know how to run an R script without the R.nvim plugin.

**Comment by b-rodrigues** [comment]
So I set up Neovim with the required R plugins, and was able to start it from a Nix Shell and run some code. There's probably a plugin or something in your .Rprofile that messes things up, but it's difficult for me to say what. Maybe you could start from a fresh config and add things one by one to identify what's wrong?

**Reply by oTheAnalyst** [reply]
I will give that a try. If I find an issue I will let you know.

I will make 1 or 2 testing branches of my neovim config, so that if I find anything it can be replicated.

**Reply by oTheAnalyst** [reply]
> So I set up Neovim with the required R plugins, and was able to start it from a Nix Shell and run some code. There's probably a plugin or something in your .Rprofile that messes things up, but it's difficult for me to say what. Maybe you could start from a fresh config and add things one by one to identify what's wrong?

could you share a gist of the config you used? I'm starting to think the issue is coming from `require("config.lazy") `

**Reply by b-rodrigues** [reply]
```
-- bootstrap lazy.nvim, LazyVim and your plugins
require("config.lazy")
require("plugins.ncm-R")
require("plugins.cmp-r")
require("plugins.r-nvim")
-- Setup language servers
require("lspconfig").r_language_server.setup({})

```

**Reply by oTheAnalyst** [reply]
Okay, so I read up on some of the documentation of R.nvim turns out if you add:
```r 
options(
  languageserver.server_capabilities =
    list(
      completionProvider = FALSE,
      completionItemResolve = FALSE
    )
)
```
To the .Rprofile it will remove the the need for having ``require("lspconfig").r_language_server.setup({})`` in init.lua it will use a neovim plugin that is already a dependency for R.nvim. 

I created ``https://github.com/VtheRtech/Vim-Config-Files/tree/minimalist`` a minimalist branch with all the nonessentials removed, and I used hooks to create a hard link to`.Rprofile` in my home directory so it switches to the new branch `.Rprofile` whenever I use `git switch`.

I followed the same steps I first posted and I'm still getting the same error. If I remove 

```r
.libPaths(c(.libPaths(), "~/Rpackages/"))
.libPaths(c(.libPaths(), "~/R/x86_64-pc-linux-gnu-library/4.3/"))
```

and go enter into the shell environment and then enter into the text editor, no package is recognized and no package can be installed and the same error is shown when I use `:RDebuginfo`

```bash
before_rns.R stdout: * checking for file ‘/home/pretender/.local/share/nvim/lazy/R.nvim/nvimcom/DESCRIPTION’ ... OK* preparing ‘nvimcom’:* checking DESCRIPTION meta-information ...
 OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘nvimcom_0.9.42.tar.gz’

Time:
  check health (async): 0.000640973
  unix setup: 2.9726e-05
  global setup: 0.001026769
  before_rns.R (async): 0.356260207
before_rns.R stderr: 
.build_packages() exit status 0
* installing to library ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’Error: ERROR: no permission to install to directory ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’
Error in do_exit(status = status) : .install_packages() exit status 1Execution halted

Why build nvimcom: Nvimcom not installed
```

**Comment by b-rodrigues** [comment]
Try adding languageserver to your environments list of packages here

https://github.com/VtheRtech/trade-union-behaviour-2020s/blob/208773bc82b6f36f49d3d12e9a5d0f2ded9343e5/default.nix#L22

Sent from Proton Mail Android

-------- Original Message --------

**Reply by oTheAnalyst** [reply]
Unfortunately, this did not work. Is it possible to install nvimcom using rix?

**Comment by b-rodrigues** [comment]
I didn't have to do anything special, nvimcom got installed automatically, but do the instructions here help?

https://nixos.wiki/wiki/Nvim-r

Since you're not using homemanager, just adding gcc to the list of packages should suffice 

Another answer might be to try to install it from Github:
https://github.com/jalvesaq/Nvim-R/tree/master/R/nvimcom

Since it's in a subfolder it might not be possible to install it like another R package, so maybe you'll need to rehost it.


-------- Original Message --------

**Reply by oTheAnalyst** [reply]
```nix 
 system_packages = builtins.attrValues {
  inherit (pkgs) pandoc R glibcLocales nix gcc;
 };
```
added the line 


```
Time:
  unix setup: 4.4083e-05
  global setup: 0.00160621
  before_rns.R (async): 0.403680811
  check health (async): 0.000748918
before_rns.R stderr: .build_packages() exit status 0* installing to library ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’Error: ERROR: no permission to install to directory ‘/nix/store/zci0ali173n6dar8w2ga3jvlalivd0s1-r-here-1.0.1/library’Error in do_exit(status = status) : .install_packages() exit status 1Execution halted

Why build nvimcom: Nvimcom not installed
before_rns.R stdout: (empty)
```
upon entering the shell with nvim the error appears 
![image](https://github.com/b-rodrigues/rix/assets/30744769/93083383-ae2f-45c5-ac18-07e87fef0e51)

This is so bizarre. Will try and install it from github next. I'm leaving out of town for a holiday and will be back  on my pc tomorrow. 

This can probably be solved by using VM or moving to a new OS (I planned on switching to nixos anyways) but I am more curious as to what the issue is.

**Comment by oTheAnalyst** [comment]
Would there be way to add:

options(
  languageserver.server_capabilities =
    list(
      completionProvider = FALSE,
      completionItemResolve = FALSE
    )
)


To a rix file so that it will build an R profile with those options? 

The culprit has to be the language server. If I disable and use the cmp plugin instead I think it should work.

**Comment by b-rodrigues** [comment]
Let's close this and open an issue instead
